# Teclov Part 1: Data Cleaning

Let's start with getting the datafiles rounds.csv and companies.txt.


In [65]:
import numpy as np
import pandas as pd 
import matplotlib.pyplot as plt
import seaborn as sns

# reading data files
# using encoding = "ISO-8859-1" to avoid pandas encoding error
rounds = pd.read_csv("rounds2.csv", encoding = "ISO-8859-1")
companies = pd.read_csv("companies.txt", sep="\t", encoding = "ISO-8859-1")


In [66]:
# Look at rounds head
print(rounds.head())

                     company_permalink  \
0                  /organization/-fame   
1               /ORGANIZATION/-QOUNTER   
2               /organization/-qounter   
3  /ORGANIZATION/-THE-ONE-OF-THEM-INC-   
4                /organization/0-6-com   

                           funding_round_permalink funding_round_type  \
0  /funding-round/9a01d05418af9f794eebff7ace91f638            venture   
1  /funding-round/22dacff496eb7acb2b901dec1dfe5633            venture   
2  /funding-round/b44fbb94153f6cdef13083530bb48030               seed   
3  /funding-round/650b8f704416801069bb178a1418776b            venture   
4  /funding-round/5727accaeaa57461bd22a9bdd945382d            venture   

  funding_round_code   funded_at  raised_amount_usd  
0                  B  05-01-2015         10000000.0  
1                  A  14-10-2014                NaN  
2                NaN  01-03-2014           700000.0  
3                  B  30-01-2014          3406878.0  
4                  A  19-03-2008      

In [67]:
# inspect the structure etc.
print(rounds.info(), "\n")
print(rounds.shape)


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 114949 entries, 0 to 114948
Data columns (total 6 columns):
company_permalink          114949 non-null object
funding_round_permalink    114949 non-null object
funding_round_type         114949 non-null object
funding_round_code         31140 non-null object
funded_at                  114949 non-null object
raised_amount_usd          94959 non-null float64
dtypes: float64(1), object(5)
memory usage: 5.3+ MB
None 

(114949, 6)


The variables funding_round_code and raised_amount_usd contain some missing values, as shown above. We'll deal with them after we're done with understanding the data - column names, primary keys of tables etc.

In [68]:
# look at companies head
companies.head()

permalink                    name  \
0                  /Organization/-Fame                   #fame   
1               /Organization/-Qounter                :Qounter   
2  /Organization/-The-One-Of-Them-Inc-  (THE) ONE of THEM,Inc.   
3                /Organization/0-6-Com                 0-6.com   
4       /Organization/004-Technologies        004 Technologies   

                        homepage_url  \
0                 http://livfame.com   
1             http://www.qounter.com   
2                http://oneofthem.jp   
3                 http://www.0-6.com   
4  http://004gmbh.de/en/004-interact   

                                       category_list     status country_code  \
0                                              Media  operating          IND   
1  Application Platforms|Real Time|Social Network...  operating          USA   
2                                  Apps|Games|Mobile  operating          NaN   
3                                        Curated Web  operating          CHN   
4                                           Software  operating          USA   

  state_code                 region           city  founded_at  
0         16                 Mumbai         Mumbai         NaN  
1         DE             DE - Other  Delaware City  04-09-2014  
2        NaN                    NaN            NaN         NaN  
3         22                Beijing        Beijing  01-01-2007  
4         IL  Springfield, Illinois      Champaign  01-01-2010

In [69]:
# companies structure
companies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 66368 entries, 0 to 66367
Data columns (total 10 columns):
permalink        66368 non-null object
name             66367 non-null object
homepage_url     61310 non-null object
category_list    63220 non-null object
status           66368 non-null object
country_code     59410 non-null object
state_code       57821 non-null object
region           58338 non-null object
city             58340 non-null object
founded_at       51147 non-null object
dtypes: object(10)
memory usage: 5.1+ MB


Ideally, the ```permalink``` column in the companies dataframe should be the unique_key of the table, having 66368 unique company names (links, or permalinks). Also, these 66368 companies should be present in the rounds file.

Let's first confirm that these 66368 permalinks (which are the URL paths of companies' websites) are not repeating in the column, i.e. they are unique.

In [71]:
# identify the unique number of permalinks in companies
len(companies.permalink.unique())

66368

Also, let's convert all the entries to lowercase (or uppercase) for uniformity.

In [72]:
# converting all permalinks to lowercase
companies['permalink'] = companies['permalink'].str.lower()
companies.head()


permalink                    name  \
0                  /organization/-fame                   #fame   
1               /organization/-qounter                :Qounter   
2  /organization/-the-one-of-them-inc-  (THE) ONE of THEM,Inc.   
3                /organization/0-6-com                 0-6.com   
4       /organization/004-technologies        004 Technologies   

                        homepage_url  \
0                 http://livfame.com   
1             http://www.qounter.com   
2                http://oneofthem.jp   
3                 http://www.0-6.com   
4  http://004gmbh.de/en/004-interact   

                                       category_list     status country_code  \
0                                              Media  operating          IND   
1  Application Platforms|Real Time|Social Network...  operating          USA   
2                                  Apps|Games|Mobile  operating          NaN   
3                                        Curated Web  operating          CHN   
4                                           Software  operating          USA   

  state_code                 region           city  founded_at  
0         16                 Mumbai         Mumbai         NaN  
1         DE             DE - Other  Delaware City  04-09-2014  
2        NaN                    NaN            NaN         NaN  
3         22                Beijing        Beijing  01-01-2007  
4         IL  Springfield, Illinois      Champaign  01-01-2010

In [73]:
# look at unique values again
len(companies.permalink.unique())

66368

Thus, there are 66368 unique companies in the table and ```permalink``` is the unique primary key. Each row represents a unique company.

Let's now check whether all of these 66368 companies are present in the rounds file, and if some extra ones are present.

In [74]:
# look at unique company names in rounds df
# note that the column name in rounds file is different (company_permalink)
len(rounds.company_permalink.unique())


90247

There seem to be 90247 unique values of ```company_permalink```, whereas we expected only 66368. May be this is because of uppercase/lowercase issues.

Let's convert the column to lowercase and look at unique values again.

In [75]:
# converting column to lowercase
rounds['company_permalink'] = rounds['company_permalink'].str.lower()
rounds.head()

company_permalink  \
0                  /organization/-fame   
1               /organization/-qounter   
2               /organization/-qounter   
3  /organization/-the-one-of-them-inc-   
4                /organization/0-6-com   

                           funding_round_permalink funding_round_type  \
0  /funding-round/9a01d05418af9f794eebff7ace91f638            venture   
1  /funding-round/22dacff496eb7acb2b901dec1dfe5633            venture   
2  /funding-round/b44fbb94153f6cdef13083530bb48030               seed   
3  /funding-round/650b8f704416801069bb178a1418776b            venture   
4  /funding-round/5727accaeaa57461bd22a9bdd945382d            venture   

  funding_round_code   funded_at  raised_amount_usd  
0                  B  05-01-2015         10000000.0  
1                  A  14-10-2014                NaN  
2                NaN  01-03-2014           700000.0  
3                  B  30-01-2014          3406878.0  
4                  A  19-03-2008          2000000.0

In [76]:
# Look at unique values again
len(rounds.company_permalink.unique())

66370

There seem to be 2 extra permalinks in the rounds file which are not present in the companies file. Let's hope that this is a data quality issue, since if this were genuine, we have two companies whose investment round details are available but their metadata (company name, sector etc.) is not available in the companies table.

Let's have a look at the company permalinks which are in the 'rounds' file but not in 'companies'.

In [77]:
# companies present in rounds file but not in (~) companies file
rounds.loc[~rounds['company_permalink'].isin(companies['permalink']), :]

company_permalink  \
29597                             /organization/e-cãŠbica   
31863        /organization/energystone-games-灵石游戏   
45176                /organization/huizuche-com-惠ç§ÿ车   
58473              /organization/magnet-tech-磁石科技   
101036  /organization/tipcat-interactive-沙èˆÿ信息ç...   
109969             /organization/weiche-tech-喂车科技   
113839                 /organization/zengame-ç¦
游科技   

                                funding_round_permalink funding_round_type  \
29597   /funding-round/8491f74869e4fe8ba9c378394f8fbdea               seed   
31863   /funding-round/b89553f3d2279c5683ae93f45a21cfe0               seed   
45176   /funding-round/8f8a32dbeeb0f831a78702f83af78a36               seed   
58473   /funding-round/8fc91fbb32bc95e97f151dd0cb4166bf               seed   
101036  /funding-round/41005928a1439cb2d706a43cb661f60f               seed   
109969  /funding-round/f74e457f838b81fa0b29649740f186d8            venture   
113839  /funding-round/6ba28fb4f3eadf5a9c6c81bc5dde6cdf               seed   

       funding_round_code   funded_at  raised_amount_usd  
29597                 NaN  01-02-2015                NaN  
31863                 NaN  09-08-2014                NaN  
45176                 NaN  18-09-2014                NaN  
58473                 NaN  16-08-2014          1625585.0  
101036                NaN  06-09-2010                NaN  
109969                  A  06-09-2015                NaN  
113839                NaN  17-07-2010                NaN

All the permalinks have weird non-English characters. Let's see whether these characters are present in the original df as well. 

In [78]:
# looking at the indices with weird characters
rounds_original = pd.read_csv("rounds2.csv", encoding = "ISO-8859-1")
rounds_original.iloc[[29597, 31863, 45176, 58473], :]

company_permalink  \
29597                       /ORGANIZATION/E-CÊBICA   
31863  /ORGANIZATION/ENERGYSTONE-GAMES-ǁµÇŸ³Æ¸¸Æˆ   
45176          /organization/huizuche-com-惠ç§ÿ车   
58473        /ORGANIZATION/MAGNET-TECH-Ç£ÇŸ³Ç§‘ÆŠ€   

                               funding_round_permalink funding_round_type  \
29597  /funding-round/8491f74869e4fe8ba9c378394f8fbdea               seed   
31863  /funding-round/b89553f3d2279c5683ae93f45a21cfe0               seed   
45176  /funding-round/8f8a32dbeeb0f831a78702f83af78a36               seed   
58473  /funding-round/8fc91fbb32bc95e97f151dd0cb4166bf               seed   

      funding_round_code   funded_at  raised_amount_usd  
29597                NaN  01-02-2015                NaN  
31863                NaN  09-08-2014                NaN  
45176                NaN  18-09-2014                NaN  
58473                NaN  16-08-2014          1625585.0

The company weird characters appear when you import the data file. To confirm whether these characters are actually present in the given data or whether python has introduced them while importing into pandas, let's have a look at the original CSV file in Excel.

Thus, this is most likely a data quality issue we have introduced while reading the data file into python. Specifically, this is most likely caused because of encoding.

First, let's try to figure out the encoding type of this file. Then we can try specifying the encoding type at the time of reading the file. The ```chardet``` library shows the encoding type of a file.

In [ ]:
# import chardet

# rawdata = open('rounds2.csv', 'rb').read()
# result = chardet.detect(rawdata)
# charenc = result['encoding']
# print(charenc)

In [ ]:
# print(result)

Now let's try telling pandas (at the time of importing) the encoding type. Here's a list of various encoding types python can handle: https://docs.python.org/2/library/codecs.html#standard-encodings.

In [ ]:
# trying different encodings
# encoding="cp1254" throws an error
# rounds_original = pd.read_csv("rounds2.csv", encoding="cp1254")
# rounds_original.iloc[[29597, 31863, 45176], :]

Apparently, pandas cannot decode "cp1254" in this case. After searching a lot on stackoverflow and Google, the best conclusion that can be drawn is that this file is encoded using multiple encoding types (may be because the ```company_permalink``` column contains names of companies in various countries, and hence various languages).

After trying various other encoding types (in vain), this answer suggested an alternate (and a more intelligent) way: https://stackoverflow.com/questions/45871731/removing-special-characters-in-a-pandas-dataframe.



In [79]:
rounds['company_permalink'] = rounds.company_permalink.str.encode('utf-8').str.decode('ascii', 'ignore')
rounds.loc[~rounds['company_permalink'].isin(companies['permalink']), :]

company_permalink  \
77                                  /organization/10north   
729                               /organization/51wofang-   
2670                              /organization/adslinked   
3166    /organization/aesthetic-everything-social-network   
3291                 /organization/affluent-attach-club-2   
4568                           /organization/allgu-outlet   
8097                             /organization/asiansbook   
8652                              /organization/atlye-gri   
9784                                  /organization/axgaz   
14311              /organization/boral-bikes-incorporated   
14798                          /organization/brasil-oznio   
14951                         /organization/bricopriv-com   
15384                                   /organization/brv   
16018                         /organization/bhner-eh-gmbh   
16624                          /organization/canal-da-pea   
16839                                 /organization/cappt   
23210                        /organization/contrato-rpido   
23518                        /organization/cop-active-ltd   
24932                          /organization/crme-ciseaux   
24933                          /organization/crme-ciseaux   
27110                         /organization/desafo-tctico   
29597                               /organization/e-cbica   
31863                    /organization/energystone-games-   
33069                              /organization/etool-io   
37562                                 /organization/freem   
37876                            /organization/frquentiel   
39460             /organization/gesto-sade-e-tecnologia-3   
42221                        /organization/grfica-en-lnea   
45176                         /organization/huizuche-com-   
46281                    /organization/ignia-bienes-races   
...                                                   ...   
58474                          /organization/magnet-tech-   
60960                    /organization/mercado-electrnico   
62172                            /organization/ming-yazlm   
63761                         /organization/monnier-frres   
65471                  /organization/mdica-santa-carmen-2   
73633                               /organization/patrofn   
78497                             /organization/prodti-cz   
79281                          /organization/przewietl-pl   
85619                              /organization/salo-vip   
85996      /organization/satlite-distribuidora-de-petrleo   
90328                               /organization/skar-is   
91932   /organization/socit-internationale-de-plantati...   
97158                         /organization/slfar-studios   
97571                          /organization/talentsigned   
99945                       /organization/the-vision-lab-   
100588                         /organization/th-gii-di-ng   
101036                  /organization/tipcat-interactive-   
104092                                /organization/tximo   
104093                            /organization/to-conejo   
105508                         /organization/vacation-bnb   
108953                             /organization/v-de-txi   
108954                 /organization/vnder-sports-network   
109968                         /organization/weiche-tech-   
109969                         /organization/weiche-tech-   
110516                      /organization/whites-holdings   
110545                       /organization/whodats-spaces   
113839                             /organization/zengame-   
114946                                 /organization/eron   
114947                               /organization/asys-2   
114948  /organization/novatiff-reklam-ve-tantm-hizmetl...   

                                funding_round_permalink   funding_round_type  \
77      /funding-round/b41ff7de932f8b6e5bbeed3966c0ed6a  equity_crowdfunding   
729     /funding-round/346b9180d276a74e0fbb2825e66c6f5b              venture   
2670    /funding-

This seems to work fine. 

Let's now look at the number of unique values in rounds dataframe again.

In [80]:
# Look at unique values again
len(rounds.company_permalink.unique())

66368

Now it makes sense - there are 66368 unique companies in both the ```rounds``` and ```companies``` dataframes. 

It is possible that a similar encoding problems are present in the companies file as well. Let's look at the companies which are present in the companies file but not in the rounds file - if these have special characters, then it is most likely because the ```companies``` file is encoded (while rounds is not).

In [81]:
# companies present in companies df but not in rounds df
companies.loc[~companies['permalink'].isin(rounds['company_permalink']), :]


permalink  \
43                               /organization/10â°north   
426                  /organization/51wofang-无忧我房   
1506                          /organization/adslinked™   
1775   /organization/aesthetic-everythingâ®-social-ne...   
1834              /organization/affluent-attachã©-club-2   
2556                        /organization/allgã¤u-outlet   
4567                         /organization/asiansbook™   
4903                           /organization/atã¶lye-gri   
5490                               /organization/axã¨gaz   
8131            /organization/borã©al-bikes-incorporated   
8385                        /organization/brasil-ozã´nio   
8477                       /organization/bricoprivã©-com   
8710                                 /organization/bräv   
9095                       /organization/bã¶hner-eh-gmbh   
9441                        /organization/canal-da-peã§a   
9569                               /organization/cappt㺠  
13126                     /organization/contrato-rã¡pido   
13286                     /organization/copã©-active-ltd   
14078                       /organization/crã¨me-ciseaux   
15306                    /organization/desafã­o-tã¡ctico   
16827                            /organization/e-cãšbica   
18197       /organization/energystone-games-灵çÿ³æ¸¸æˆ   
18926                           /organization/etool-io┠  
21578                              /organization/freemå   
21769                         /organization/frã©quentiel   
22711          /organization/gesto-saãºde-e-tecnologia-3   
24327                   /organization/grã¡fica-en-lã­nea   
26139               /organization/huizuche-com-惠租车   
26818                 /organization/ignia-bienes-raã­ces   
26892                        /organization/ikigã¼nde-com   
...                                                  ...   
31693                            /organization/lawpã dã­   
33892             /organization/magnet-tech-磁çÿ³ç§‘暀   
35353                 /organization/mercado-electrã´nico   
36039                       /organization/ming-yazä±lä±m   
36917                      /organization/monnier-frã¨res   
37951               /organization/mã©dica-santa-carmen-2   
42529                            /organization/patrofä°n   
45338                          /organization/prodä›ti-cz   
45760                       /organization/przeå›wietl-pl   
49431                           /organization/salã£o-vip   
49635  /organization/satã©lite-distribuidora-de-petrã...   
52055                            /organization/skarã¸-is   
52994  /organization/sociã©tã©-internationale-de-plan...   
56005                      /organization/sã³lfar-studios   
56251                      /organization/talentsigned™   
57710                    /organization/the-vision-lab-â®   
58099              /organization/thế-giới-di-äá»™ng   
58344  /organization/tipcat-interactive-沙舟信息ç...   
60102                              /organization/tã¡ximo   
60103                          /organization/tã­o-conejo   
60981                      /organization/vacation-bnb™   
62884                         /organization/vã¡-de-tã¡xi   
62885               /organization/vã¼nder-sports-network   
63486             /organization/weiche-tech-喂车科暀   
63811                   /organization/white’s-holdings   
63833                    /organization/whodat’s-spaces   
65778                 /organization/zengame-ç¦
游科暀   
66365                               /organization/ãeron   
66366                             /organization/ã”asys-2   
66367  /organization/ä°novatiff-reklam-ve-tanä±tä±m-h...   

                                                    name  \
43                                             10°North   
426                                51wofang 无忧我房   
1506                                        AdsLinked™   
1775               Aesthetic Everything® Social Network   
1834                      

Thus, the ```companies``` df also contains special characters. Let's treat those as well.

In [82]:
# remove encoding from companies df
companies['permalink'] = companies.permalink.str.encode('utf-8').str.decode('ascii', 'ignore')


Let's now look at the companies present in the companies df but not in rounds df - ideally there should be none.

In [83]:
# companies present in companies df but not in rounds df
companies.loc[~companies['permalink'].isin(rounds['company_permalink']), :]


Empty DataFrame
Columns: [permalink, name, homepage_url, category_list, status, country_code, state_code, region, city, founded_at]
Index: []

Thus, the encoding issue seems resolved now. Let's write these (clean) dataframes into separate files so we don't have to worry about encoding problems again.

In [84]:
# write rounds file
rounds.to_csv("rounds_clean.csv", sep=',', index=False)

# write companies file
companies.to_csv("companies_clean.csv", sep='\t', index=False)